src/common/DOMWorld.ts

```diff
@@ -464,6 +473,82 @@ export class DOMWorld {
     return queryHandler.waitFor(this, updatedSelector, options);
   }
 
+  /**
+   * @internal
+   */
+  async addBindingToContext(name: string) {
+    const expression = helper.pageBindingInitString('internal', name);
+    try {
+      const context = await this.executionContext();
+      await context._client.send('Runtime.addBinding', {
+        name,
+        executionContextId: context._contextId,
+      });
+      await context.evaluate(expression);
+    } catch (error) {
+      // When the page is navigated, the promise is rejected.
+      // We will try again in the new execution context.
+      if (error.message.includes('Execution context was destroyed')) return;
+      // We could have tried to evaluate in a context which was already
+      // destroyed.
+      if (error.message.includes('Cannot find context with specified id'))
+        return;
+      debugError(error);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  async addBinding(name: string, puppeteerFunction: Function): Promise<void> {
+    if (this._ctxBindings.has(name))
+      throw new Error(
+        `Failed to add page binding with name ${name}: window['${name}'] already exists!`
+      );
+    this._ctxBindings.set(name, puppeteerFunction);
+    await this.addBindingToContext(name);
+  }
+
+  /**
+   * @internal
+   */
+  hasBinding(name: string): boolean {
+    return this._ctxBindings.has(name);
+  }
+
+  private async _onBindingCalled(
+    event: Protocol.Runtime.BindingCalledEvent
+  ): Promise<void> {
+    let jsonPayload = null;
+    try {
+      jsonPayload = JSON.parse(event.payload);
+    } catch {
+      // Binding must have been called before initializing the wrapper.
+      return;
+    }
+    const { type, name, seq, args } = jsonPayload;
+    if (type !== 'internal' || !this._ctxBindings.has(name)) return;
+    if (!this._hasContext()) return;
+    const context = await this.executionContext();
+    if (context._contextId !== event.executionContextId) return;
+    try {
+      const result = await this._ctxBindings.get(name)(...args);
+      await context.evaluate(deliverResult, name, seq, result);
+    } catch (error) {
+      // The WaitTask may already have been resolved by timing out, or the
+      // exection context may have been destroyed.
+      // In both caes, the promises above are rejected with a protocol error.
+      // We can safely ignores these, as the WaitTask is re-installed in
+      // the next execution context if needed.
+      if (error.message.includes('Protocol error')) return;
+      debugError(error);
+    }
+    function deliverResult(name: string, seq: number, result: unknown): void {
+      window[name].callbacks.get(seq).resolve(result);
+      window[name].callbacks.delete(seq);
```

Comment:
`window` -> `globalThis` like elsewhere in the PR? That way we support worker contexts as well. Even if those don't need to be supported in this code path, it's more consistent with other code in this PR

Reply:
Done.